ipa-core/src/protocol/ipa_prf/shuffle/mod.rs

```diff
@@ -0,0 +1,167 @@
+use self::base::shuffle;
+use super::boolean_ops::{expand_array_in_place, extract_from_shared_array};
+use crate::{
+    error::Error,
+    ff::{
+        boolean::Boolean,
+        boolean_array::{BA112, BA64},
+        ArrayAccess, CustomArray, Expand, Field,
+    },
+    protocol::context::{UpgradableContext, UpgradedContext},
+    report::OprfReport,
+    secret_sharing::{
+        replicated::{semi_honest::AdditiveShare, ReplicatedSecretSharing},
+        SharedValue,
+    },
+};
+
+pub mod base;
+
+#[tracing::instrument(name = "shuffle_inputs", skip_all)]
+async fn shuffle_inputs<C, BK, TV, TS>(
+    ctx: C,
+    input: Vec<OprfReport<BK, TV, TS>>,
+) -> Result<Vec<OprfReport<BK, TV, TS>>, Error>
+where
+    C: UpgradableContext,
+    C::UpgradedContext<Boolean>: UpgradedContext<Boolean, Share = AdditiveShare<Boolean>>,
+    BK: SharedValue + CustomArray<Element = Boolean> + Field,
+    TV: SharedValue + CustomArray<Element = Boolean> + Field,
+    TS: SharedValue + CustomArray<Element = Boolean> + Field,
+    for<'a> &'a AdditiveShare<TS>: IntoIterator<Item = AdditiveShare<Boolean>>,
+    for<'a> &'a AdditiveShare<TV>: IntoIterator<Item = AdditiveShare<Boolean>>,
+    for<'a> &'a AdditiveShare<BK>: IntoIterator<Item = AdditiveShare<Boolean>>,
+    for<'a> <&'a AdditiveShare<TV> as IntoIterator>::IntoIter: Send,
+    for<'a> <&'a AdditiveShare<TS> as IntoIterator>::IntoIter: Send,
+{
+    let shuffle_input: Vec<AdditiveShare<BA112>> = input
+        .into_iter()
+        .map(|item| oprfreport_to_shuffle_input::<BA112, BK, TV, TS>(&item))
+        .collect::<Vec<_>>();
+
+    let shuffled = shuffle(ctx, shuffle_input).await?;
+
+    Ok(shuffled
+        .into_iter()
+        .map(|item| shuffled_to_oprfreport(&item))
+        .collect::<Vec<_>>())
+}
+
+// This function converts OprfReport to an AdditiveShare needed for shuffle protocol
+pub fn oprfreport_to_shuffle_input<YS, BK, TV, TS>(
+    input: &OprfReport<BK, TV, TS>,
+) -> AdditiveShare<YS>
+where
+    YS: CustomArray<Element = <BA112 as CustomArray>::Element> + SharedValue,
+    BK: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
+    TV: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
+    TS: SharedValue + ArrayAccess<Output = Boolean> + Expand<Input = Boolean>,
+{
+    let (mut y_left, mut y_right) = (YS::ZERO, YS::ZERO);
+    expand_array_in_place(&mut y_left, input.match_key.left(), 0);
```

Comment:
was trying to avoid multiple allocations for y. Is there a way to pass mutable tuples and change their values inside a function?

Reply:
worked!